crates/pathfinder/src/update.rs

```diff
@@ -0,0 +1,170 @@
+            }
+            UpdateResult::ReqwestError(e) if e.is_decode() || e.is_body() || e.is_builder() => {
+                // More severe errors, probably indicating something is wrong with our setup.
+                // Set to warn and not error because this update checking is a non-critical feature.
+                tracing::warn!(error=%e, "Error checking Github for new releases")
+            }
+            UpdateResult::ReqwestError(e) => {
+                // Less severe errors, includes transient connection errors and timeouts; does not warrant
+                // a high log level.
+                tracing::trace!(error=%e, "Error checking Github for new releases")
+            }
+            UpdateResult::Other(e) => {
+                // Other, unexpected errors.
+                tracing::warn!(error=%e, "Error checking Github for new releases");
+            }
+        }
+
+        // Display latest release info if it is newer than this application.
+        if let Some(github) = latest_gh_version.as_ref() {
+            if github > &local_version {
+                tracing::warn!(release=%github, "New pathfinder release available! Please consider updating your node!")
+            }
+        }
+
+        // Only poll every 10 minutes.
+        tokio::time::sleep(std::time::Duration::from_secs(60 * 10)).await;
+    }
+}
+
+#[derive(Debug)]
+struct Release {
+    version: String,
+    /// Optional because its possible to have an invalid string ETag header.
+    etag: Option<String>,
+}
+
+#[derive(Debug)]
+enum UpdateResult {
+    /// A new latest release.
+    Update(Release),
+    /// No new releases since last query (status code 304).
+    NotModified,
+    ReqwestError(reqwest::Error),
+    Other(anyhow::Error),
+}
+
+/// Fetches the latest pathfinder [Release] from Github using their REST API.
+///
+/// The [IF_NONE_MATCH](reqwest::header::IF_NONE_MATCH) header options is set to
+/// the `etag` parameter to prevent Github from sending redundant information. The
+/// resulting 304 status code is mapped to [UpdateResult::NotModified].
+async fn fetch_latest_github_release(
+    client: &reqwest::Client,
+    etag: &Option<String>,
+) -> UpdateResult {
+    use reqwest::StatusCode;
+    use reqwest::Url;
+
+    const USER_AGENT: &str = concat!(
+        env!("CARGO_PKG_NAME"),
+        "/",
+        env!("VERGEN_GIT_SEMVER_LIGHTWEIGHT")
+    );
+    let url = Url::parse("https://api.github.com/repos/eqlabs/pathfinder/releases/latest").unwrap();
+
+    let mut request = client
+        .get(url)
+        .timeout(std::time::Duration::from_secs(300))
+        // https://docs.github.com/en/rest/overview/resources-in-the-rest-api#current-version
+        .header(reqwest::header::ACCEPT, "application/vnd.github.v3+json")
+        // https://docs.github.com/en/rest/overview/resources-in-the-rest-api#user-agent-required
+        .header(reqwest::header::USER_AGENT, USER_AGENT);
```

Comment:
Minor nit: can't we just set the user-agent on the client itself (instead of setting it explicitly for each request)?

https://docs.rs/reqwest/latest/reqwest/struct.ClientBuilder.html#method.user_agent